lib/elixir/lib/list.ex

```diff
@@ -467,62 +467,44 @@ defmodule List do
       iex> List.ascii_printable?('abc' ++ [0], 2)
       true
 
-  Improper lists are not printable, even if made only of ascii characters:
+  Last element in an improper list renders them not printable, even if made only of ASCII characters:
 
       iex> List.ascii_printable?('abc' ++ ?d)
       false
 
+      iex> List.ascii_printable?('abc' ++ ?d, 3)
+      true
+
   """
   @doc since: "1.6.0"
-  def ascii_printable?(list, counter \\ :infinity)
-
-  def ascii_printable?(_, 0) do
-    true
-  end
-
-  def ascii_printable?([char | rest], counter)
-      when is_integer(char) and char >= 32 and char <= 126 do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\n | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\r | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\t | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\v | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
+  @spec ascii_printable?(maybe_improper_list, limit) :: boolean
+        when limit: :infinity | non_neg_integer
+  def ascii_printable?(list, limit \\ :infinity)
+      when is_list(list) and (limit == :infinity or (is_integer(limit) and limit >= 0)) do
+    ascii_printable_with_limit?(list, limit)
   end
 
-  def ascii_printable?([?\b | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
+  @compile {:inline, ascii_printable_with_limit?: 2}
```

Comment:
inlining recursive functions usually is not very beneficial

Reply:
in general, it's highly specific from case to case, but overall, it generally doesn't provide huge benefits. You'd need to measure it

Reply:
thank you